examples/wmt/train.py

```diff
@@ -0,0 +1,638 @@
+  metrics = {
+      'loss': loss,
+      'accuracy': acc,
+      'denominator': weight_sum,
+  }
+  metrics = jax.lax.psum(metrics, axis_name='batch')
+  return metrics
+
+
+def train_step(optimizer,
+               batch,
+               learning_rate_fn,
+               label_smoothing=0.0,
+               use_bfloat16=False,
+               dropout_rng=None):
+  """Perform a single training step."""
+  train_keys = ['inputs', 'targets',
+                'inputs_position', 'targets_position',
+                'inputs_segmentation', 'targets_segmentation']
+  (inputs, targets,
+   inputs_positions, targets_positions,
+   inputs_segmentation, targets_segmentation) = [
+       batch.get(k, None) for k in train_keys]
+
+  weights = jnp.where(targets > 0, 1, 0).astype(jnp.float32)
+
+  # We handle PRNG splitting inside the top pmap to improve efficiency.
+  dropout_rng, new_dropout_rng = random.split(dropout_rng)
+
+  def loss_fn(model):
+    """loss function used for training."""
+    with nn.stochastic(dropout_rng):
+      logits = model(
+          inputs,
+          targets,
+          use_bfloat16=use_bfloat16,
+          inputs_positions=inputs_positions,
+          targets_positions=targets_positions,
+          inputs_segmentation=inputs_segmentation,
+          targets_segmentation=targets_segmentation,
+          train=True,
+          cache=None)
+
+    loss, weight_sum = compute_weighted_cross_entropy(logits, targets, weights,
+                                                      label_smoothing)
+    mean_loss = loss / weight_sum
+    return mean_loss, logits
+
+  step = optimizer.state.step
+  lr = learning_rate_fn(step)
+  grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+  (_, logits), grad = grad_fn(optimizer.target)
+  grad = jax.lax.pmean(grad, 'batch')
+  new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+  metrics = compute_metrics(logits, targets, weights)
+  metrics['learning_rate'] = lr
+
+  return new_optimizer, metrics, new_dropout_rng
+
+
+def eval_step(model, batch, label_smoothing=0.0, use_bfloat16=False):
+  """Calculate evaluation metrics on a batch."""
+  inputs, targets = batch['inputs'], batch['targets']
+  weights = jnp.where(targets > 0, 1.0, 0.0)
+  logits = model(inputs, targets, use_bfloat16=use_bfloat16, train=False,
+                 cache=None)
+  return compute_metrics(logits, targets, weights, label_smoothing)
+
+
+def predict_step(inputs, model, cache, eos_token, max_decode_len):
+  """Predict translation with fast decoding beam search on a batch."""
+  batch_size = inputs.shape[0]
+  beam_size = 4
+
+  # Prepare transformer fast-decoder call for beam search:
+  # for beam search, we need to set up our decoder model
+  # to handle a batch size equal to batch_size * beam_size,
+  # where each batch item's data is expanded in-place rather
+  # than tiled.
+  # i.e. if we denote each batch element subtensor as el[n]:
+  # [el0, el1, el2] --> beamsize=2 --> [el0,el0,el1,el1,el2,el2]
```

Comment:
```suggestion
  # Prepare transformer fast-decoder call for beam search: for beam search, we
  # need to set up our decoder model to handle a batch size equal to 
  # batch_size * beam_size, where each batch item's data is expanded in-place
  # rather than tiled.
  # i.e. if we denote each batch element subtensor as el[n]:
  # [el0, el1, el2] --> beamsize=2 --> [el0,el0,el1,el1,el2,el2]
```